src/app/application/services/data-source/index.ts

```diff
@@ -1,36 +1,97 @@
+          if (doc.hidden) {
+            await new Promise((resolve) => {
+              doc.addEventListener('visibilitychange', resolve, { once: true })
+            })
+          }
+        }
+        let res
+        try {
+          res = cb(self)
+          // if we aren't polling then immediately close after calling
+          if (typeof self.configuration.interval === 'undefined') {
+            self.close()
+          }
+          // only increase iterations if we didn't error
+          iterations++
+          // return the result
+          yield res
+        } catch (e) {
+          // if retry is configured await it before entering the loop again to
+          // try again
+          // TODO(jc): we should probably pass through attempts and maybe other
+          // things here
+          const retry = self.configuration?.retry?.(e)
+          if (typeof retry?.then === 'function') {
+            // make sure we never mistakenly retry sooner than 1s
+            await Promise.all([retry, new Promise(resolve => setTimeout(resolve, 1000))])
+            attempts++
+          } else {
+            throw e
+          }
+        }
+      }
+    }, config)
+  }
+}
+export type Source = ReturnType<typeof getSource>
+
+// its fine to not wait for an unfocussed tab for promise returning sources
+const source = getSource(new (class extends EventTarget {hidden = false})())
 
 export const create: Creator = (src, router) => {
   const [path, query] = src.split('?')
   const queryParams = new URLSearchParams(query)
   // use the router to find which function to call
   const route = router.match(path)
-  const _source = new CallableEventSource(async function * (this: CallableEventSource) {
-    while (true) {
-      this.readyState = 1
-      // `.route` here is the function call to the 'source' i.e. the Promise
-      // returning call that can be polled, in our case right now the HTTP
-      // calls but in the future could also be 'listeners' on localStorage, or
-      // 'listeners' on a session
-      yield route.route({
-        ...{
-          offset: parseInt(queryParams.get('offset') || '0'),
-          size: parseInt(queryParams.get('size') || '0'),
-          page: parseInt(queryParams.get('page') || '0'),
-          search: queryParams.get('search') || '',
-        },
-        ...route.params,
-      }, this)
-      if (!isClosed(this)) {
-        // right now any polling has a 5s interval, we currently just hardcode
-        // here but if/when we use this more this should be a user setting
-        // that we can save/retrieve from localStorage
-        await new Promise(resolve => setTimeout(resolve, 5000))
-      }
+  const params = {
+    ...{
+      offset: parseInt(queryParams.get('offset') || '0'),
+      size: parseInt(queryParams.get('size') || '0'),
+      page: parseInt(queryParams.get('page') || '0'),
+      search: queryParams.get('search') || '',
+    },
+    ...route.params,
+  }
```

Comment:
note that every single data source has these available in its `params` argument (both before and after this PR)